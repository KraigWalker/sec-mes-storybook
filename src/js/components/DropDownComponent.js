import React from 'react';

class DropDownComoonent extends React.Component {
    constructor(props) {
        super(props);
        this.returnMenuItem = this.returnMenuItem.bind(this);
    };
    returnMenuItem() {
        if(this.props.id === 'accounts'){
            let items = [];  
            _.map(this.props.accounts.accounts,(account)=>{
                items.push(<option key = {account} value ={account}>{account}</option>
            );
            })
            return items;
           
        } else {
            let items = [];  
            _.map(this.props.subjects.subjects, (subject) => {
                console.log('sub:',subject);
                items.push(<option key = {subject.key} value={subject.value} >{subject.value}</option>);
            },false);
            return items;
        }
       
    }
    render() {
        return( <div>
        <select onChange = {this.props.selectSubject} name = "select" id={this.props.id}>
        {this.returnMenuItem()}
    </select><br/><br/><br/>
    </div>);
    }
}
export default DropDownComoonent;