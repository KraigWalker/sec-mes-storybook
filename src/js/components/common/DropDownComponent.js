import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import MessageSubjectEntity from '../../entities/MessageSubjectEntity';
import { connect } from 'react-redux';
import { getMessageSubjects, getAccounts, sendMessageData } from '../../actions/AppActions';
import CalloutComponent from './CalloutComponent';
class DropDownComponent extends React.Component {
    constructor(props) {
        super(props);
        this.returnMenuItem = this.returnMenuItem.bind(this);
        this.overlayclick = this.overlayclick.bind(this);
        this.showList = this.showList.bind(this);
        this.setDropDrownValue = this.setDropDrownValue.bind(this);
        this.onBlur = this.onBlur.bind(this);
        this.state = {
            list: false,
            Text: this.props.selectedValue,
            isDropDownDisabled: false,
        }
    };
    componentWillMount() {
        //Will remove after service call testing of accounts
        // if (!this.props.messagesubjects.fetched && !this.props.messageaccounts.fetched) {
        // this.props.dispatch(getMessageSubjects());
        // this.props.dispatch(getAccounts());
        // }
        // if (this.props.isFromDraftOrReply) {
        //     this.props.selectSubject(this.props.selectedValue, this.props.id);
        // }
    }
    returnMenuItem() {
        let setDropDrownValue;
        let items = [];
        switch (true) {
            case (!this.props.isFromDraftOrReply && this.props.id === 'accounts'):
                items.push(<li className="c-dropdown__value" id="No specific account" key='No specific account' value='No specific account' onClick={e => this.setDropDrownValue(e, 'accounts')}>No specific account</li>);
                _.map(this.props.accounts.accounts, (account) => {
                    items.push(<li className="c-dropdown__value" id={account.name} key={account.name} value={account.name} onClick={e => this.setDropDrownValue(e, account)}>{account.name}</li>
                    );
                                   /* To display account name along with account number use the below pattern
                    <li className="c-dropdown__value" key={account} value={account} onClick={this.setDropDrownValue}><span className="c-dropdown__value__account">{account}</span><span className="c-dropdown__value__number">1234</span></li>
                */
                });
                break;
            case (!this.props.isFromDraftOrReply && this.props.id === 'subjects'):
                _.map(this.props.subjects.subjects, (subject) => {
                    items.push(<li className="c-dropdown__value" key={subject.key} id={subject.value} onClick={e => this.setDropDrownValue(e, subject)}>{subject.value}</li>);
                }, false);
                break;
            case (this.props.isFromDraftOrReply && this.props.id === 'accounts'):
                items.push(<li className="c-dropdown__value" id="No specific account" key='No specific account' value='No specific account' onClick={e => this.setDropDrownValue(e, 'accounts')}>No specific account</li>);
                _.map(this.props.messageaccounts.accounts, (account) => {
                    items.push(<li className="c-dropdown__value" id={account.name} key={account.name} value={account.name} onClick={e => this.setDropDrownValue(e, account)}>{account.name}</li>
                    );
                               /* To display account name along with account number use the below pattern
                    <li className="c-dropdown__value" key={account} value={account} onClick={this.setDropDrownValue}><span className="c-dropdown__value__account">{account}</span><span className="c-dropdown__value__number">1234</span></li>
                */
                })
                break;
            case (this.props.isFromDraftOrReply && this.props.id === 'subjects'):
                _.map(this.props.messagesubjects.subjects, (subject) => {
                    items.push(<li className="c-dropdown__value" key={subject.key} id={subject.value} onClick={e => this.setDropDrownValue(e, subject)}>{subject.value}</li>
                    );
                })
                break;
        }
        return items;
    }
    overlayclick() {
        this.setState({
            list: false,
        });

    }
    showList() {
        if (this.state.list === false) {
            this.setState({ list: true });
        }
        else {
            this.setState({ list: false });
        }
    }
    setDropDrownValue(e, typeOfData) {
        this.setState({
            Text: e.target.textContent,
            list: false,
        });
        this.props.selectSubject(e.target.textContent, this.props.id, typeOfData);

    }
    onBlur() {
        this.setState({
            list: false,
        });
    }
    render() {
        const overlayClassName = cx({
            'c-overlay overlay__custom--zindex': true,
            'overlay__show': this.state.list,

        });
        return (
            <div>
                <div>
                    <button id="ddlText" aria-label={`${this.props.accessID} ${this.state.Text}`} className="c-field__input c-field__input--select c-dropdown u-cursor-pointer" onClick={this.showList}>{this.state.Text}</button>
                    {this.state.list && <div ref="overlay" className={overlayClassName} onClick={this.overlayclick} ></div>}
                    {this.state.list && <ul className="c-dropdown__list u-cursor-pointer" onBlur={this.onBlur}>{this.returnMenuItem()}</ul>}
                </div>
                {this.props.showAccountError ? <CalloutComponent dClass='callout callout__error callout__inline-error' paraText='Please select a subject for your message' /> : ''}
                {this.props.showSubjectError ? <CalloutComponent dClass='callout callout__error callout__inline-error' paraText='Pleae select an option for which account your enquiry relates to. If itâ€™s a general enquiry, choose `General enquiry`' /> : ''}
            </div>
        );
    }
}

/**
 * Maps the state of the component to the state of the redux store
 * @param {object} state. State of the application
 */
const mapState = (state) => {
    return {
        messagesubjects: state.subjects,
        messageaccounts: state.accounts,
    }
};

export default connect(mapState)(DropDownComponent);