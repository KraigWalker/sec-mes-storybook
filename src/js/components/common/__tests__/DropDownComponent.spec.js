import React from 'react';
import { shallow, mount } from 'enzyme';
import { DropDownComponent } from '../DropDownComponent';

describe("Dropdown snapshot", () => {
    const selectSubject = jest.fn();
    let props = {
        content: {
            back: 'Back',
        },
        accessID: 'Subject',
        messages: [],
        successModal: false,
        selectedAccountValue: null,
        subjects: [],
        subjectErrors: false,
        showSubjectError: false,
        showAccountError: false,
        accounts: [],
        id: 'subjects',
        selectSubject: selectSubject,
        popupState: jest.fn()

    };
    let component = shallow(<DropDownComponent {...props} />);
    component.setState({ showErrorModal: true });
    it('should match to snapshot', () => {
        expect(component).toMatchSnapshot();
    });
    
});
