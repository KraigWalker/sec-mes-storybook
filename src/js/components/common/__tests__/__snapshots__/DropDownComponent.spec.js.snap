// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown snapshot should match to snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DropDownComponent
    accessID="Subject"
    accounts={Array []}
    content={
      Object {
        "back": "Back",
      }
    }
    id="subjects"
    messages={Array []}
    popupState={[MockFunction]}
    selectSubject={[MockFunction]}
    selectedAccountValue={null}
    showAccountError={false}
    showSubjectError={false}
    subjectErrors={false}
    subjects={Array []}
    successModal={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Column
          className="u-padding-0"
          ld={9}
          md={9}
          sm={24}
          xs={24}
        >
          <TextBody>
            <Select
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "label": undefined,
                    "value": undefined,
                  },
                ]
              }
            />
          </TextBody>
        </Column>,
        "",
        "",
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <TextBody>
            <Select
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "label": undefined,
                    "value": undefined,
                  },
                ]
              }
            />
          </TextBody>,
          "className": "u-padding-0",
          "ld": 9,
          "md": 9,
          "sm": 24,
          "xs": 24,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Select
              defaultValue={undefined}
              disabled={false}
              id={undefined}
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "label": undefined,
                    "value": undefined,
                  },
                ]
              }
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultValue": undefined,
              "disabled": false,
              "id": undefined,
              "onChange": [Function],
              "options": Array [
                Object {
                  "label": undefined,
                  "value": undefined,
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      "",
      "",
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Column
            className="u-padding-0"
            ld={9}
            md={9}
            sm={24}
            xs={24}
          >
            <TextBody>
              <Select
                defaultValue={undefined}
                disabled={false}
                id={undefined}
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": undefined,
                      "value": undefined,
                    },
                  ]
                }
              />
            </TextBody>
          </Column>,
          "",
          "",
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <TextBody>
              <Select
                defaultValue={undefined}
                disabled={false}
                id={undefined}
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": undefined,
                      "value": undefined,
                    },
                  ]
                }
              />
            </TextBody>,
            "className": "u-padding-0",
            "ld": 9,
            "md": 9,
            "sm": 24,
            "xs": 24,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Select
                defaultValue={undefined}
                disabled={false}
                id={undefined}
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": undefined,
                      "value": undefined,
                    },
                  ]
                }
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "defaultValue": undefined,
                "disabled": false,
                "id": undefined,
                "onChange": [Function],
                "options": Array [
                  Object {
                    "label": undefined,
                    "value": undefined,
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        "",
        "",
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenFourAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
